### DISCUSSION TITLE
Software

### DISCUSSION THREAD

**Discussion by KenWillmott** [KenWillmott/HD6303EVB#12] [General]
I have a view of where I'd like to take the platform software. It's a layered model, designed to facilitate middle and high level programming on this fairly primitive CPU.

1. MLL - Machine Language Layer. Low level arbitrary codes written directly in 68xx assembly language.
2. MDP - Machine Design Patterns. A collection of standard code modules and rules, to extend the basic processor instruction set to solve common functional needs. Implemented as assembler macros, dedicated memory assignments, and coding rules.
3. VML - Virtual Machine layer. An emulation of a processor with enhanced characteristics vs. the 68xx. Can be interpreted, threaded or compiled.
4. HLL - High Level Language. A translator that produces VML object code from text based source code. Either an existing, or new language.

Elsewhere, it is common to see a HLL compile for an MLL, whether resident or cross-compiling is used. Either that, or an HLL runs via an interpreter that uses something like the VML. The four layer scheme spreads out the difficulty of translation between layers, and allows for redundant support. For example, once a VML or MDP exists, different HLL may use the same underlying VML or MDP. This makes the addition of new languages easier. Also the MDP and VML may be maintained and improved, independently of the HLL's.

It's the MDP that is really new here. The idea came from the CDP1802 literature. The CDP1802 has no call stack to speak of. So the literature suggests a code design pattern to follow, which emulates a subroutine call, using a standard register assignment and code call/return sequence. The 68xx is not nearly so limited, but I believe there are aspects which can benefit from the same approach. There are several areas in which this could be useful. Most of the 68xx have only a single 16 bit index register, IX. So when more than one pointer is required, things become awkward, standard solutions are slow and may have side effects. For example, the X register can be pushed onto the stack. But then, if there is a branch after that, it can not precede the corresponding X register pull.

In practice, the MDP elements can be coded as assembler macros, but it's not sufficient in itself, those operate in conjunction with a set of rules and guidelines that integrate them smoothly with MLL. It is expected that most code will be a mixture of MLL and MDP, when that is the case, the MLL is not specifically restricted in operations, but must respect the MDP rules and paradigm.

**Comment by TheEccentricGenius** [comment]
OK. I'll think about all this and see if I can help.

A lot work to implement all this.....

**Comment by KenWillmott** [comment]
I think, back in the day, coders were in too much hurry to make things work. That's how we ended up with a decade of interpretive BASIC. My theory is that extra time spent building some solid foundations, will make things easier and easier as the system grows... instead of hitting a wall because of tools that are not forward looking. This is both retroactive computing, "what if" fun, and also a proof of concept that might spill over to contemporary hobby computing.

**Reply by TheEccentricGenius** [reply]
Some software or assembler problem....I have the kbug6303.asm file from the github repo, I try to compile it for a test run and I get a list of mnemonic errors....Did I do  something wrong?

I have just used the commands in the way you recommended in the other discussion.
Also I have the Cygwin version as I have no Linux setup currently.

**Comment by KenWillmott** [comment]
That version is written in Motorola assembler syntax. In the beginning of development, I used a Motorola cross assembler that I compiled from source. Unfortunately, I lost track of where I got it from! Later, I realized that I would need a macro assembler, and couldn't find a Motorola syntax compatible one. ASxxxx looked good to me, so I adopted it, there are others like dasm. I converted the monitor to ASxxxx syntax, and uploaded it with a new name:
[https://github.com/KenWillmott/HD6303EVB/blob/main/kbug6303-103-asx.asm](url)
The Hitachi examples and all other code is also written in ASxxxx syntax.

The ASxxxx page:
[https://shop-pdp.net/ashtml/](url)
I'm running it in the command line interface on Windows 11. I edit using Eclipse for embedded, I found it almost understands the syntax, at least highlights comments in a different colour which is nice. To complete that, a specialization would have to be added to Eclipse, which I'm definitely not getting into right now. But it's functional and useful.

**Reply by TheEccentricGenius** [reply]
Sorry, I had a older version of the repo....Thank you for solution.

I don't have Eclipse either....

**Comment by KenWillmott** [comment]
You don't really need Eclipse, for a while I was just using Windows notepad. :)

**Reply by TheEccentricGenius** [reply]
LOL. I do too.

**Comment by TheEccentricGenius** [comment]
Did you find a way to get binary out of the assembler?

If you didn't, I'll make a S-Record converter, so I can use my Arduino EEPROM Programmer to burn S-Records to my EEPROM.

**Comment by TheEccentricGenius** [comment]
Also do I have to modify KBug for HD6303RP....My quick look only told me it should work, but I am no expert.

**Comment by KenWillmott** [comment]
No, as far as I can tell, the Baldwin assembler has no option to produce raw binaries. Kbug should work without big changes. The place to double check, is the short initialization section at the START label, controlling the UART settings. Those have to match your CPU and board. Currently, it's configured for the simplest settings - internal baud derived from 7.37MHz CPU crystal, no E stretching on peripheral.

**Reply by TheEccentricGenius** [reply]
I checked that bit appears to all check out OK.

**Comment by KenWillmott** [comment]
On the subject of burning binaries to EEPROM, there's another back burner idea I'm working on - using the board itself as a programmer. The bus connector can (or can be made) to access EEPROM if the signals are "turned around" and used as a controller. With the MCU reset line held active, wire jumpers could connect (for example) an Arduino Mega R3 I/O pins with the address and data bus. Then, a simple loader program running on the Mega could install the entire monitor in the EEPROM. It would be the only time you would normally ever need to do it, because firmware updates are possible over serial (just not written yet...).

The EVB has one catch with that, it uses the direct read and write lines from the processor, they are not decoded externally. But it can still work if headers bring those two signals out somewhere. Again, you should only need to do it once. If I recall correctly the 6303RP doesn't bring those out by default, so they have to be decoded by a NAND circuit that gates read and write with system clock E. That probably means you could do everything from the bus connector for most 6303RP designs.

**Reply by TheEccentricGenius** [reply]
Interesting....I'll see what I can do when I get around to it.

Unfortunately the S9 in the S-Record has me confused....Does information go where the S1 record specifies or S9?

**Comment by KenWillmott** [comment]
All the payload data is in the S9 records, so the simplified parser in the MIKBUG, MINIBUG and such Motorola monitors ignores all other S record types. Kbug is based on code from those, so it's the same.

**Comment by KaelWebb** [comment]
Your platform software vision sounds fascinating!

**Comment by JamariRamirez** [comment]
Love the layered model approach you're taking – it's like building a solid software sandwich with each layer having its unique flavor. The MLL and MDP tiers seem like the foundation, adding that low-level magic. And the VML and HLL layers? They're like the icing on the cake, bringing enhanced features and translating ideas into reality. If anyone needs to explore [custom software development](https://cosource.com.au/software-development/), this layered concept could be a great blueprint! Keep up the innovative work! 👏🚀

**Comment by KenWillmott** [comment]
Thanks! I'd like to explore the approach in more detail, but I had to put all software aside to get the boards made and tested. I promise that I will do that, for now I can only speak generally. One example, the 6800 has only one index register and two accumulators, which forces you to stop and think about something even as simple as a "loop and count" like a "for" loop. Also the most efficient codes use zero page memory, only 256 bytes... actually less with the 6303 because the control registers use space there. So the MDP would deal with that. Basically the codes would solve those problems in the usual ways. But the idea is to standardize those ways, minimize collisions between unit implementations. They would be available usually as macros. Along with that, uniform expectations of register persistence (the problem manifests itself when the programmer has to stop and think carefully about what register contents need to be preserved during the execution of a block). Anyone who has done this has also run up against the management of status flags as well, this can lead to a lot of back and forth reading of the instruction set summary. You can never completely avoid having to know about it, but the idea is to find patterns that will avoid the need to drill down that far, maybe 98% of the time.

The HLL part, it could actually be any language but I thought (perhaps ambitiously) that it would be fun to improve on Forth by adding an operator stack so as to make it a normal infix syntax, not the postfix (reverse Polish notation) syntax. The threaded implementation of Forth seems to have worked very well and so is worth consideration.

I think I can start some serious software work once the processor and game boards reach a stable non-beta version. The hardware is eating up all my time now.
